crates/nu-protocol/src/value/mod.rs

```diff
@@ -370,10 +373,32 @@ impl Value {
         }
     }
 
+    /// Returns this `Value` converted to a `String` or an error if it cannot be converted
+    ///
+    /// Only the following `Value` cases will return an `Ok` result:
+    /// - `Int`
+    /// - `Float`
+    /// - `String`
+    /// - `Binary` (only if valid utf-8)
+    /// - `Date`
```

Comment:
Thanks for making this clear across the different methods!

Should we doctest this?
(preferrably probably the inverse but that would be unwieldy to read)

Reply:
It's quite a simple function so I'm not sure. Maybe the doctests could go in an `# Examples` section to make it more readable?

Reply:
I think the more unfortunate situation would be if someone adds another coercion and doesn't document it but that is kind of harder to write concisely. (Having just the given ones as one line asserts would be easy as examples)

Reply:
I was thinking of doing something like this:
```
/// ```
/// # use nu_protocol::{
/// #     ast::{CellPath, RangeInclusion},
/// #     engine::Closure,
/// #     Range, Record, ShellError, Span, Value
/// # };
/// # use chrono::DateTime;
/// #
/// # let values: Vec<Value> = vec![
/// #     Value::test_bool(false),
/// #     Value::test_int(0),
/// #     Value::test_filesize(0),
/// #     Value::test_duration(0),
/// #     Value::test_date(DateTime::UNIX_EPOCH.into()),
/// #     Value::test_range(Range {
/// #         from: Value::test_nothing(),
/// #         incr: Value::test_nothing(),
/// #         to: Value::test_nothing(),
/// #         inclusion: RangeInclusion::Inclusive,
/// #     }),
/// #     Value::test_float(0.0),
/// #     Value::test_string(String::new()),
/// #     Value::test_record(Record::new()),
/// #     // Value::test_lazy_record(),
/// #     Value::test_list(Vec::new()),
/// #     Value::test_block(0),
/// #     Value::test_closure(Closure {
/// #         block_id: 0,
/// #         captures: Vec::new(),
/// #     }),
/// #     Value::test_nothing(),
/// #     Value::error(
/// #         ShellError::NushellFailed { msg: String::new() },
/// #         Span::test_data(),
/// #     ),
/// #     Value::test_binary(Vec::new()),
/// #     Value::test_cell_path(CellPath {
/// #         members: Vec::new(),
/// #     }),
/// #     // Value::test_custom_value(0),
/// # ];
/// #
/// # for val in &values {
/// assert_eq!(matches!(val, Value::Float { .. } | Value::Int { .. }), val.coerce_float().is_ok())
/// # };
/// ```
```

But this would have to be copy-pasted for each function. Or, we add a public function somewhere that returns a `Vec<Value>` containing one of each `Value` case (perhaps it could be `#[doc(hidden)]`).

Reply:
It's a mouthful 😄 

(as we already have the `Value::test_...` in the general API it probably wouldn't be too bad to have a sample generating function)

but I don't want to block those changes on this. It will cause churn with a lot of PRs anyways.